File: solver/src/main/java/will/sudoku/solver/detectors/BoxLineReductionDetector.kt
package will.sudoku.solver.detectors

import will.sudoku.solver.Board
import will.sudoku.solver.Coord
import will.sudoku.solver.HintGenerator
import will.sudoku.solver.TechniqueDetector

/**
 * Detects box/line reductions — candidates restricted to one box in a row or column.
 */
class BoxLineReductionDetector : TechniqueDetector {

    override val technique = HintGenerator.Technique.BOX_LINE_REDUCTION

    override fun detect(board: Board): HintGenerator.Hint? {
        // Check rows for box/line reduction
        for (row in 0..8) {
            for (value in 1..9) {
                val cells = mutableListOf<Coord>()
                for (col in 0..8) {
                    val coord = Coord(row, col)
                    if (!board.isConfirmed(coord) &&
                        value in board.candidateValues(coord)
                    ) {
                        cells.add(coord)
                    }
                }
                if (cells.size in 2..3) {
                    val boxes = cells.map { Pair(it.row / 3, it.col / 3) }.toSet()
                    if (boxes.size == 1) {
                        val (boxRow, boxCol) = boxes.first()
                        for (r in boxRow * 3 until (boxRow + 1) * 3) {
                            if (r == row) continue
                            for (c in boxCol * 3 until (boxCol + 1) * 3) {
                                val coord = Coord(r, c)
                                if (!board.isConfirmed(coord) &&
                                    value in board.candidateValues(coord)
                                ) {
                                    return HintGenerator.Hint(
                                        coord = coord,
                                        value = value,
                                        technique = technique,
                                        explanation = "Value $value in row ${row + 1} " +
                                                "is restricted to box (${boxRow + 1},${boxCol + 1}). " +
                                                "Eliminate $value from other rows in this box."
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }

        // Check columns for box/line reduction
        for (col in 0..8) {
            for (value in 1..9) {
                val cells = mutableListOf<Coord>()
                for (row in 0..8) {
                    val coord = Coord(row, col)
                    if (!board.isConfirmed(coord) &&
                        value in board.candidateValues(coord)
                    ) {
                        cells.add(coord)
                    }
                }
                if (cells.size in 2..3) {
                    val boxes = cells.map { Pair(it.row / 3, it.col / 3) }.toSet()
                    if (boxes.size == 1) {
                        val (boxRow, boxCol) = boxes.first()
                        for (c in boxCol * 3 until (boxCol + 1) * 3) {
                            if (c == col) continue
                            for (r in boxRow * 3 until (boxRow + 1) * 3) {
                                val coord = Coord(r, c)
                                if (!board.isConfirmed(coord) &&
                                    value in board.candidateValues(coord)
                                ) {
                                    return HintGenerator.Hint(
                                        coord = coord,
                                        value = value,
                                        technique = technique,
                                        explanation = "Value $value in column ${col + 1} " +
                                                "is restricted to box (${boxRow + 1},${boxCol + 1}). " +
                                                "Eliminate $value from other columns in this box."
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }

        return null
    }
}

File: packages/solver/src/detectors/BoxLineReductionDetector.ts
import type { Board } from '../Board'
import { Coord } from '../Coord'
import type { Hint, TechniqueDetector } from '../HintTypes'
import { Technique } from '../HintTypes'

/**
 * Detects box/line reductions — candidates restricted to one box in a row or column.
 *
 * When a candidate value in a row (or column) is confined to a single box,
 * that value can be eliminated from the rest of that box.
 */
export class BoxLineReductionDetector implements TechniqueDetector {
    readonly technique = Technique.BOX_LINE_REDUCTION

    detect(board: Board): Hint | null {
        // Check rows for box/line reduction
        for (let row = 0; row < 9; row++) {
            for (let value = 1; value <= 9; value++) {
                const cells: Coord[] = []
                for (let col = 0; col < 9; col++) {
                    const coord = Coord.all[row * 9 + col]
                    if (!board.isConfirmed(coord) &&
                        board.candidateValues(coord).includes(value)) {
                        cells.push(coord)
                    }
                }
                if (cells.length >= 2 && cells.length <= 3) {
                    const boxes = new Set(cells.map(c => `${Math.floor(c.row / 3)},${Math.floor(c.col / 3)}`))
                    if (boxes.size === 1) {
                        const [boxRow, boxCol] = boxes.values().next().value!.split(',').map(Number)
                        for (let r = boxRow * 3; r < boxRow * 3 + 3; r++) {
                            if (r === row) continue
                            for (let c = boxCol * 3; c < boxCol * 3 + 3; c++) {
                                const coord = Coord.all[r * 9 + c]
                                if (!board.isConfirmed(coord) &&
                                    board.candidateValues(coord).includes(value)) {
                                    return {
                                        coord,
                                        value,
                                        technique: Technique.BOX_LINE_REDUCTION,
                                        explanation: `Value ${value} in row ${row + 1} ` +
                                            `is restricted to box (${boxRow + 1},${boxCol + 1}). ` +
                                            `Eliminate ${value} from other rows in this box.`
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        // Check columns for box/line reduction
        for (let col = 0; col < 9; col++) {
            for (let value = 1; value <= 9; value++) {
                const cells: Coord[] = []
                for (let row = 0; row < 9; row++) {
                    const coord = Coord.all[row * 9 + col]
                    if (!board.isConfirmed(coord) &&
                        board.candidateValues(coord).includes(value)) {
                        cells.push(coord)
                    }
                }
                if (cells.length >= 2 && cells.length <= 3) {
                    const boxes = new Set(cells.map(c => `${Math.floor(c.row / 3)},${Math.floor(c.col / 3)}`))
                    if (boxes.size === 1) {
                        const [boxRow, boxCol] = boxes.values().next().value!.split(',').map(Number)
                        for (let c = boxCol * 3; c < boxCol * 3 + 3; c++) {
                            if (c === col) continue
                            for (let r = boxRow * 3; r < boxRow * 3 + 3; r++) {
                                const coord = Coord.all[r * 9 + c]
                                if (!board.isConfirmed(coord) &&
                                    board.candidateValues(coord).includes(value)) {
                                    return {
                                        coord,
                                        value,
                                        technique: Technique.BOX_LINE_REDUCTION,
                                        explanation: `Value ${value} in column ${col + 1} ` +
                                            `is restricted to box (${boxRow + 1},${boxCol + 1}). ` +
                                            `Eliminate ${value} from other columns in this box.`
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        return null
    }
}
